[dynamo][DebugMode] make ModTracker a no-op in compiled regions

pianpwk opened the pull request:
ModTracker causes a graph break in compiled regions: https://github.com/pytorch/pytorch/issues/169995, so this makes it a no-op by introducing the `torch._dynamo.eval_frame._is_in_compiled_region()` check.

The next PR in the stack makes nn.Module tracking work for DebugMode by introducing an interpreter.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170126
* #170125
* __->__ #170124



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

pianpwk committed:
silence ModTracker in compiled regions

[ghstack-poisoned]

yushangdi reviewed (commented):
LGTM, we should also get a review from the dynamo folks

tugsbayasgalan reviewed (approved):
Accepting to unblock, but pls consider making is_compiling flag extensive enough to cover this case as well.

pianpwk committed:
Update on "[dynamo][DebugMode] make ModTracker a no-op in compiled regions"

ModTracker causes a graph break in compiled regions: https://github.com/pytorch/pytorch/issues/169995, so this makes it a no-op by introducing the `torch._dynamo.eval_frame._is_in_compiled_region()` check.

The next PR in the stack makes nn.Module tracking work for DebugMode by introducing an interpreter.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 reviewed (commented):
Why does torch.compiler.is_compiling not work here?

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -203,6 +205,24 @@ def get_example_inputs(key: str) -> list[Any]:
 
 
+@contextlib.contextmanager
+def _set_in_compiled_region():
```
Can we add more documentation on how this differs from _compiling_state_context()/torch.compiler.is_compiling()?

pianpwk committed:
Update on "[dynamo][DebugMode] make ModTracker a no-op in compiled regions"

ModTracker causes a graph break in compiled regions: https://github.com/pytorch/pytorch/issues/169995, so this makes it a no-op by introducing the `torch._dynamo.eval_frame._is_in_compiled_region()` check.

The next PR in the stack makes nn.Module tracking work for DebugMode by introducing an interpreter.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
